In [ ]:
# Paso 2: Cargar y explorar.

In [9]:
import pandas as pd

# 1. Carga del CSV desde la URL pública especificada
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
# 2. Obtener la información estructural del Dataframe con: 891 filas entre tipos de datos enteros, de punto flotante y strings.
df.info()
# 3. Mostrar las primeras 5 filas del DataFrame
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [16]:
# Obtención a través de pandas en vez de resta manual
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [12]:
# 4. Generar estadísticas
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [14]:
# 5. Contando categorías
df['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [15]:
df['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

# Entregable: Carga y conclusiones del análisis exploratorio de datos.
Haciendo uso de las respuestas que brindaron el comando df.info() podemos identificar cuáles son los campos que comprenden más o menos valores nulos que no serían de utilidad para responder nuestras preguntas objetivo y que podrían representar un sesgo dependiendo del porcentaje en el cuál se encuentren estas variables.

RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
| Nº | Columna  | Valores No Nulos | Tipo de Dato |
|----|----------|------------------|--------------|
| 1  | Survived | 891 non-null     | int64        |
| 5  | Age      | 714 non-null     | float64      |
| 10 | Cabin    | 204 non-null     | str          |
| 11 | Embarked | 889 non-null     | str          |   

Dentro de la tabla se encuentra las cifras de la columna Non-Null Count que verifica cuántas filas y de qué campos no se encuentran con valores nulos. Para el campo de 'Survived' cada una de las entradas de este comprenden valores válidos y aptos para sacar una hipótesis a través de operaciones. Pero los campos de 'Age' 'Cabin' 'Embarked' representan perdidas de datos del 177, 687 y 2 filas comprendiendo una perdida de 19.87%, 77.10% y 0.22% respectivamente (calculado a través de num de nulos entre el num total de filas multiplicado por 100 para obtener el porcentaje). Dentro de un 5% es considerado perdida marginal por lo que podría eliminar esas filas sin introducir un sesgo que descalifique mis conclusiones o imputarlas con medidas de tendencia central ('Embarked'), por otro lado entre 5% y 30% Es necesario la imputación de datos con los metodos anteriormente mencionados o utilizando modelos predictivos ('Age'). Por último, el 77.10% de perdida de información no puede ser solucionado con imputación de datos independientemente de la herramienta que quieras utilizar, donde deberíamos de eliminar la columna porque no es de utilidad para el estudio.

Pasemos a la fase de limpieza.


### Justificación Técnica de las Decisiones de Limpieza

1. **Uso de la Mediana en `Age` en lugar de la Media:**
   Al observar los datos estadísticos previos con `df.describe()`, la distribución de la edad presenta un sesgo debido a la presencia de niños y adultos mayores, desvíandose de una distribución estrictamente normal. La media aritmética es altamente sensible a los valores extremos (outliers), lo que alteraría la estructura real de la población si la usáramos para rellenar los vacíos. La mediana representa el percentil 50 ($28$ años), garantizando una imputación robusta y central no afectada por los extremos.

2. **Nulos de `Embarked` imputados con la moda:**

    Al ser un campo que mantiene un tipo de dato en strings cabe dentro del a categoría de datos nominales donde las respuestas de 'S' 'C' o 'Q' representan puertos y no tienen un sentido matemático. Por lo que la única forma l´+ogica de rellenar los datos faltantes sin alterar la tendencia es usar el puerto donde se subió la gran mayoría de personas. La moda, que en este caso es 'S' por ser el puerto de origen con 644 registros.

3. **Eliminación de la Columna `Cabin`:**
   Con un $77.10\%$ de valores nulos, cualquier intento de imputación tradicional introduciría un ruido masivo y sesgos artificiales en los análisis posteriores, perdiendo utilidad predictiva o explicativa confiable.

In [17]:
# Paso 3: Limpiar
# 1. Imputación de nulos en Age (~20%) con la mediana
mediana_edad = df['Age'].median()
df['Age'] = df['Age'].fillna(mediana_edad)

# 2. Imputación de nulos en Embarked (~2 filas) con la moda
moda_embarque = df['Embarked'].mode()[0]
df['Embarked'] = df['Embarked'].fillna(moda_embarque)

# 3. Eliminación de la columna Cabin por exceso de nulos (~77%)
df = df.drop(columns='Cabin')

# 4. Verificación de nulos residuales en las columnas de trabajo
print("Conteo de nulos remanentes por columna:")
print(df.isnull().sum())

# 5. Verificación de filas duplicadas en el dataset
print(f"\nCantidad de registros completamente duplicados: {df.duplicated().sum()}")

Conteo de nulos remanentes por columna:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Cantidad de registros completamente duplicados: 0


In [18]:
# Paso 4: Transformar (Feature engineering)
# 1. FamilySize: Creación del tamaño de la familia (Hermanos/Cónyuges + Padres/Hijos + Pasajero)
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

# 2. IsAlone: Variable binaria (1 si viaja completamente solo, 0 si viaja acompañado)
df['IsAlone'] = 0
df.loc[df['FamilySize'] == 1, 'IsAlone'] = 1

# 3. AgeGroup: Segmentación de la edad en cuatro rangos ordinales definidos
bins = [0, 12, 18, 60, 100]
labels = ['Niño', 'Adolescente', 'Adulto', 'Mayor']
df['AgeGroup'] = pd.cut(df['Age'], bins=bins, labels=labels, include_lowest=True)

# 4. Title: Extracción del título social del nombre mediante expresiones regulares
df['Title'] = df['Name'].str.extract(r' ([A-Za-z]+)\.')

# Mostrar las frecuencias de los títulos extraídos
print("Distribución de Títulos (Title) extraídos de la columna Name:")
print(df['Title'].value_counts())

Distribución de Títulos (Title) extraídos de la columna Name:
Title
Mr          517
Miss        182
Mrs         125
Master       40
Dr            7
Rev           6
Major         2
Mlle          2
Col           2
Don           1
Mme           1
Ms            1
Lady          1
Sir           1
Capt          1
Countess      1
Jonkheer      1
Name: count, dtype: int64


In [21]:
# ==============================================================================
# Paso 5 — Analizar con groupby
# ==============================================================================

print("=== 1. TASA DE SUPERVIVENCIA POR GÉNERO (Sex) ===")
tasa_sex = df.groupby('Sex')['Survived'].mean() * 100
print(tasa_sex.map('{:.2f}%'.format))
print("\n" + "="*50 + "\n")

print("=== 2. TASA DE SUPERVIVENCIA POR CLASE DEL BOLETO (Pclass) ===")
tasa_pclass = df.groupby('Pclass')['Survived'].mean() * 100
print(tasa_pclass.map('{:.2f}%'.format))
print("\n" + "="*50 + "\n")

print("=== 3. TASA DE SUPERVIVENCIA POR GRUPO DE EDAD (AgeGroup) ===")
tasa_age = df.groupby('AgeGroup', observed=False)['Survived'].mean() * 100
print(tasa_age.map('{:.2f}%'.format))
print("\n" + "="*50 + "\n")

print("=== 4. COMBINACIÓN MULTINIVEL: GÉNERO Y CLASE ===")
tabla_sex_pclass = df.groupby(['Sex', 'Pclass'])['Survived'].agg(['mean', 'count'])
tabla_sex_pclass['mean'] = tabla_sex_pclass['mean'] * 100
tabla_sex_pclass.columns = ['Tasa_Supervivencia (%)', 'Total_Pasajeros']
# Usamos display() para mantener el formato de tabla interactiva de Jupyter
display(tabla_sex_pclass.style.format({'Tasa_Supervivencia (%)': '{:.2f}%'}))
print("\n" + "="*50 + "\n")

print("=== 5. ANÁLISIS MULTIVARIADO AVANZADO (Hipótesis) ===")
analisis_multivariado = df.groupby(['Sex', 'Pclass', 'AgeGroup'], observed=False)['Survived'].agg(['mean', 'count'])
analisis_multivariado.columns = ['Tasa_Supervivencia (%)', 'Total_Pasajeros']
analisis_multivariado['Tasa_Supervivencia (%)'] = analisis_multivariado['Tasa_Supervivencia (%)'] * 100

# Ordenar de forma descendente y aplicar formato de tabla estilizada
analisis_final = analisis_multivariado.sort_values(by='Tasa_Supervivencia (%)', ascending=False)
display(analisis_final.style.format({'Tasa_Supervivencia (%)': '{:.2f}%'}, na_rep='0.00%'))

=== 1. TASA DE SUPERVIVENCIA POR GÉNERO (Sex) ===
Sex
female    74.20%
male      18.89%
Name: Survived, dtype: str


=== 2. TASA DE SUPERVIVENCIA POR CLASE DEL BOLETO (Pclass) ===
Pclass
1    62.96%
2    47.28%
3    24.24%
Name: Survived, dtype: str


=== 3. TASA DE SUPERVIVENCIA POR GRUPO DE EDAD (AgeGroup) ===
AgeGroup
Niño           57.97%
Adolescente    42.86%
Adulto         36.58%
Mayor          22.73%
Name: Survived, dtype: str


=== 4. COMBINACIÓN MULTINIVEL: GÉNERO Y CLASE ===




=== 5. ANÁLISIS MULTIVARIADO AVANZADO (Hipótesis) ===


In [22]:
# Paso 7: Exportación de Hallazgos

In [23]:
import os

# Definir la ruta relativa hacia la carpeta 'data' desde la carpeta 'notebook'
ruta_salida = os.path.join('..', 'data', 'titanic_clean.csv')

# Exportar el DataFrame a CSV sin el índice jerárquico de filas
df.to_csv(ruta_salida, index=False)

print(f"Dataset exportado exitosamente en: {os.path.abspath(ruta_salida)}")

Dataset exportado exitosamente en: C:\Users\veris\aponte.dev\VSProjects\EDA-Dataset-Titanic-con-pandas\data\titanic_clean.csv
